src/server/hset_family.cc

```diff
@@ -1141,9 +1141,24 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
   auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
   OpResult<StringVec> result = cntx->transaction->ScheduleSingleHopT(std::move(cb));
   if (result) {
-    rb->SendStringArr(*result);
+    if ((result->size() == 1) && (args.size() == 1))
+      rb->SendBulkString(result->front());
+    else {
+      if ((rb->IsResp3()) && (args.size() == 3)) {  // has withvalues
+        rb->StartArray(result->size() / 2);
+        for (unsigned int i = 0; i < result->size() / 2; ++i) {
+          StringVec sv{(*result)[i * 2], (*result)[i * 2 + 1]};
+          rb->SendStringArr(sv);
+        }
+      } else {
+        rb->SendStringArr(*result);
```

Comment:
I may be missing something, but I think that instead of checking for Resp3 you could just pass `CollectionType::MAP` to `StartStringArr()`?

Reply:
yes! i didn't know this ... this is much better solution. fixed.